compose/ui/ui/src/webCommonW3C/kotlin/androidx/compose/ui/platform/accessibility/ComposeWebSemanticsListener.kt

```diff
@@ -0,0 +1,363 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.compose.ui.platform.accessibility
+
+import androidx.collection.MutableScatterMap
+import androidx.compose.ui.currentTimeMillis
+import androidx.compose.ui.geometry.Offset
+import androidx.compose.ui.platform.PlatformContext
+import androidx.compose.ui.semantics.Role
+import androidx.compose.ui.semantics.SemanticsActions
+import androidx.compose.ui.semantics.SemanticsConfiguration
+import androidx.compose.ui.semantics.SemanticsNode
+import androidx.compose.ui.semantics.SemanticsOwner
+import androidx.compose.ui.semantics.SemanticsProperties
+import kotlinx.browser.document
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.channels.BufferOverflow
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.flow.debounce
+import kotlinx.coroutines.flow.merge
+import kotlinx.coroutines.flow.receiveAsFlow
+import kotlinx.coroutines.flow.sample
+import kotlinx.coroutines.launch
+import org.w3c.dom.HTMLElement
+
+internal class ComposeWebSemanticsListener(
+    val coroutineScope: CoroutineScope,
+    val webSemanticsRoot: HTMLElement,
+) : PlatformContext.SemanticsOwnerListener {
+
+    private val invalidationChannel =
+        Channel<Unit>(1, onBufferOverflow = BufferOverflow.DROP_LATEST)
+
+    init {
+        coroutineScope.launch {
+            var lastSyncTimeMs = currentTimeMillis()
+            val invalidationFlow = invalidationChannel.receiveAsFlow()
+
+            @OptIn(FlowPreview::class)
+            merge(
+                invalidationFlow.sample(1000),
+                invalidationFlow.debounce(100)
+            ).collect {
+                val currentTime = currentTimeMillis()
+                if (currentTime - lastSyncTimeMs >= 100) {
+                    lastSyncTimeMs = currentTime
+                    syncSemanticsWithWebA11Y()
+                }
+            }
+        }
+    }
+
+    private var semanticsOwner: SemanticsOwner? = null
+
+    override fun onSemanticsOwnerAppended(semanticsOwner: SemanticsOwner) {
+        this.semanticsOwner = semanticsOwner
+    }
+
+    override fun onSemanticsOwnerRemoved(semanticsOwner: SemanticsOwner) {
+        if (semanticsOwner == this.semanticsOwner) {
+            this.semanticsOwner = null
+        }
+    }
+
+    override fun onSemanticsChange(semanticsOwner: SemanticsOwner) {
+        invalidationChannel.trySend(Unit)
```

Comment:
isn't better to use: `coroutineScope.launch { invalidationChannel.send(Unit) }`?

Reply:
It's okay for us to not send if a channel is full. We don't need a guaranteed delivery in this case.